Jetpack Backup: Add Backup retention management setting

Initsogar opened the pull request:
#### Proposed Changes

* Add a new `Days of backups saved` (backup retention period) setting in Jetpack Cloud, under Settings page. For this PR, it will be available only for A12s (dev & stage environment). The idea is to start the review process of this new functionality.
* This iteration includes:
  * Allow to update the setting only to a number of days that doesn't require a storage upgrade.
  * Fetch and update the customer-defined retention period to our backend.
  * Confirmation dialog when attempting to update to a lower retention period.

#### Demo

https://user-images.githubusercontent.com/1488641/218225112-ce96783e-d079-447b-8df2-24d9f54666fd.mov

↪ [Demo v1 (deprecated)](https://user-images.githubusercontent.com/1488641/217998763-cc89187c-9ce8-4536-a8fb-785c153ab76c.mov)


#### Screenshots

| Description | Screenshot |
|---|---|
| Setting page | ![image](https://user-images.githubusercontent.com/1488641/218224377-49cd64f7-bc66-4748-a0fa-999049c8e638.png) |
| Confirmation dialog | ![image](https://user-images.githubusercontent.com/1488641/218224478-25866771-8b23-49ee-9094-441fa6a1960a.png) |
| Tooltip | ![image](https://user-images.githubusercontent.com/1488641/218224599-6b25c3cb-d23b-46f2-a851-c97d086dfe33.png) |
| Success notice | ![image](https://user-images.githubusercontent.com/1488641/218224332-d387d0ab-e73e-4d3b-9377-54dba3f57fab.png) |
| Error notice | ![image](https://user-images.githubusercontent.com/1488641/218000224-10eb567e-2f3c-4e94-8613-c1f2947d8328.png) |




#### Testing Instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Start a Jetpack Cloud live branch
* Navigate to Settings and wait for the new setting to load.
* The `Update settings` button will be disabled when the current plan or an option that requires an upgrade are selected.
* You could try to update the days of backups saved. The way to re-confirm it was properly saved is refreshing the page.

Notes:
* If you haven't configured this setting before, the current retention period associated to your plan will be preselected. Otherwise, your current configuration will be preselected.

#### Pre-merge Checklist

<!--
Complete applicable items on this checklist **before** merging into trunk. Inapplicable items can be left unchecked.

Both the PR author and reviewer are responsible for ensuring the checklist is completed.
-->

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [x] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [x] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-ajp-p2)?
<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Initsogar committed:
Add useStorageText hook

This hook will be helpful to convert bytes to human readable size formats.

Initsogar committed:
Add BackupRetentionManagement component

Initsogar committed:
Add BackupRetentionManagement on settings page

Initsogar committed:
Use getBackupCurrentSiteSize() to get site size

Initsogar committed:
Fix initial rendering and add loading placeholder

Initsogar committed:
Fix retention options type

Initsogar committed:
Use grid instead of flexbox

Initsogar committed:
Update spaceNeeded types

Initsogar committed:
Add success and error notice when updating backup retention

Initsogar committed:
Add update settings logic

Initsogar committed:
Enable feature only for dev/staging environment

Initsogar committed:
Fix to customer retention period validation

Initsogar committed:
Refactor SCSS to avoid too much nesting

Initsogar committed:
Remove unused useMemo

Initsogar committed:
Fix spaceNeeded type

Initsogar committed:
Fix confirmation dialog max-width

Initsogar committed:
Add feature flag in jetpack-cloud-horizon

grbicsanja commented:
This is looking great! 🙌
Can we please have the confirmation notice say:
`You have successfully changed the time period of saved backups to X days`

rustamshaikh commented on client/components/backup-retention-management/hooks.tsx:
```diff
@@ -0,0 +1,73 @@
+import getBackupCurrentSiteSize from 'calypso/state/rewind/selectors/get-backup-current-site-size';
+import getRewindBytesAvailable from 'calypso/state/rewind/selectors/get-rewind-bytes-available';
+import { getSelectedSiteId } from 'calypso/state/ui/selectors';
+import { RetentionRadioOptionType } from './consts';
```
```suggestion
import type { RetentionRadioOptionType } from './consts';
```
Since we are importing type.

rustamshaikh commented on client/components/backup-retention-management/hooks.tsx:
```diff
@@ -0,0 +1,73 @@
+	const lastBackupSize = useSelector(
+		( state ) => getBackupCurrentSiteSize( state, siteId ) as number
+	);
+	return useMemo( () => lastBackupSize + lastBackupSize * 0.25, [ lastBackupSize ] );
```
`0.25` - Maybe we can define it as a constant?

Initsogar replied:
Added it to the constants.ts file, thanks!

rustamshaikh commented on client/components/backup-retention-management/consts.tsx:
```diff
@@ -0,0 +1,17 @@
+	RETENTION_DAYS_365: 365,
+} as const;
+
+export interface RetentionRadioOptionType {
```
Maybe we can move this type `RetentionRadioOptionType` to a 'type only' file say `types.ts`

Initsogar replied:
I agree. Just moved it to a new types.ts file.

grbicsanja commented:
The info icon and tooltip is missing. [See design here](https://www.figma.com/file/zEkRPQwNT9aykpCyJvrSt5/Allow-customers-to-define-their-own-backup-retention?node-id=3147%3A38368&t=u5jSmMRkhRo6u8Sm-4).
<img width="463" alt="Screenshot 2023-02-10 at 13 26 14" src="https://user-images.githubusercontent.com/90194580/218092255-195b99d8-06f1-4a1e-9516-a229b432cfbd.png">

grbicsanja commented:
I see the warning prompt even when I'm choosing a longer time period. It should only be shown when I am reducing the time and loosing saved backups.

Here in this screenshot, I was moving from 7 days to a year.
<img width="1440" alt="Screenshot 2023-02-10 at 13 27 26" src="https://user-images.githubusercontent.com/90194580/218092696-62247f2c-f78f-41b4-a0ae-1b37244a4c32.png">

rustamshaikh commented on client/components/backup-retention-management/retention-options/retention-options-control.tsx:
```diff
@@ -0,0 +1,42 @@
+import { useEffect, useState, useCallback } from '@wordpress/element';
+import React from 'react';
+import { RetentionRadioOptionType } from '../consts';
```
`import type` - Same w.r.t import type and separating constants from types.

rustamshaikh commented on client/components/backup-storage-space/hooks.tsx:
```diff
@@ -111,3 +111,49 @@ export const useDaysOfBackupsSavedText = (
+
+					return translate( '%(storageInBytes).1fTB', {
+						args: { storageInBytes: unitAmount },
+						comment: 'Must use unit abbreviation; describes an storage amounts (e.g., 1.5TB)',
```
```suggestion
						comment: 'Must use unit abbreviation; describes an storage amounts with 1 decimal point (e.g., 1.5TB)',
```
Nitpicking, but it may be helpful to translators.

Initsogar replied:
Done! Thanks for the suggestion.

rustamshaikh commented on client/jetpack-cloud/sections/settings/controller.tsx:
```diff
@@ -22,19 +24,24 @@ export const advancedCredentials: PageJS.Callback = ( context, next ) => {
-			loadingComponent={ <AdvancedCredentialsLoadingPlaceholder /> }
-		/>
+		<>
+			{ config.isEnabled( 'jetpack/backup-retention-settings' ) ? (
```
Great!

rustamshaikh reviewed (commented):
Great work @Initsogar 👍 Tested it from Jetpack Cloud live and works as expected. Left some feedback. Let me know what you think.

Initsogar committed:
Update successNotice copy, including the days

Initsogar committed:
Show confirmation only for a lower retention days

This commit also includes:
* Disable submit buttons when we started the retention update API call.
* Adjust spinner size so it can fit in the submission button without altering the size.

Initsogar committed:
Add info tooltip

Initsogar committed:
Remove margin-bottom on setting title

Initsogar committed:
Update translation comments

Initsogar committed:
Add `types.ts` file to handle type definitions

Initsogar committed:
Refactor constants

* Renaming `consts.tsx` to `constants.ts`
* Adding storage estimation additional buffer and learn more link to constants file.
* Renaming RetentionOptions to RETENTION_OPTIONS to comply with constants naming conventions.

Initsogar committed:
Adjust spacing between elements

Initsogar commented:
Thanks @grbicsanja and @rustamshaikh, just addressed all your comments and suggestions:
* Added tooltip. Also included it on the PR description, so we can double check the wording there.
* Updated success notice text to `You have successfully changed the time period of saved backups to X days`
* Update confirmation dialog to only be required when updating to a lower retention period
* Adjusted translations comments
* Refactored types and constants
* Adjusted spacing between elements

Initsogar committed:
Add hooks tests

Initsogar committed:
Remove useTranslate test mock

rustamshaikh reviewed (approved):
Verified the fixes and it looks great to me 🎉 - Thanks @Initsogar for adding more tests 👍

tyxla commented on client/components/backup-storage-space/hooks.tsx:
```diff
@@ -111,3 +111,50 @@ export const useDaysOfBackupsSavedText = (
+export const useStorageText = ( storageInBytes: number ): TranslateResult | string => {
+	const translate = useTranslate();
+
+	return useMemo( () => {
```
Is the memoization here worth the effort? What's the performance impact with and without it?

tyxla commented on client/components/backup-retention-management/types.ts:
```diff
@@ -0,0 +1,10 @@
+	spaceNeeded: TranslateResult | string; // 700MB
+	upgradeRequired: boolean;
+	isCurrentPlan: boolean;
+	value: number; // 7
```
Is this comment required or incomplete perhaps?

```suggestion
	value: number;
```

Initsogar replied:
Incomplete 🥲. The 7 was actually an example.

Either way, [we refactored it a bit](https://github.com/Automattic/wp-calypso/pull/73303/files#diff-71e240ea9b6bd13fea335936ce41b9d60fc06babcbc766ce1ed40cde9f4905e5), including a `RetentionPeriod` type instead of `number`, so we can enforce to only use the valid values (7, 30, 120, 365 for now).

Addressed in #73303

tyxla commented on client/components/backup-retention-management/loading/index.tsx:
```diff
@@ -0,0 +1,10 @@
+import * as React from 'react';
```
Should we import only the `React.FC` type if we really need that only?

Initsogar replied:
Fixed!

tyxla commented on client/components/backup-retention-management/loading/style.scss:
```diff
@@ -0,0 +1,16 @@
+.backup-retention-management__loading {
+	&.loading {
```
If this adds an extra `loading` class, why do we need the `__loading` suffix for `backup-retention-management__loading` ?

tyxla commented on client/components/backup-retention-management/info-tooltip/index.tsx:
```diff
@@ -0,0 +1,69 @@
+				{ translate( '{{a}}Learn more…{{/a}}', {
+					components: {
+						a: (
+							<a
```
Could use `<ExternalLink />` here.

tyxla commented on client/components/backup-retention-management/info-tooltip/index.tsx:
```diff
@@ -0,0 +1,69 @@
+						<Gridicon icon="cross" size={ 18 } />
+					</Button>
+				</p>
+				<hr />
```
Using `hr`s is a suboptimal way to lay out the presentation. Could be a better bet to add an extra wrapper and use CSS.

tyxla commented on client/components/backup-retention-management/info-tooltip/index.tsx:
```diff
@@ -0,0 +1,69 @@
+import { Button, Gridicon } from '@automattic/components';
+import { useTranslate } from 'i18n-calypso';
+import * as React from 'react';
```
Do we need all of `react` to be imported?

Initsogar replied:
You are right! Just fixed it.

tyxla commented on client/components/backup-retention-management/retention-options/retention-option-card.tsx:
```diff
@@ -0,0 +1,75 @@
+import classnames from 'classnames';
+import { useTranslate } from 'i18n-calypso';
+import React from 'react';
```
Same on importing all of `react`

Initsogar replied:
Fixed!

tyxla commented on client/components/backup-retention-management/retention-options/retention-option-card.tsx:
```diff
@@ -0,0 +1,75 @@
+
+interface RetentionOptionCardProps {
+	label: string;
+	spaceNeeded: TranslateResult | string;
```
Won't the `TranslateResult` include `string` already?

Also, it's an odd prop name - it sounds like a boolean but it's not. Thoughts?

Initsogar replied:
> Won't the TranslateResult include string already?

It seems [it is actually defined as a React.ReactChild](https://github.com/Automattic/wp-calypso/blob/trunk/packages/i18n-calypso/types/index.d.ts#L59) and Typescript doesn't like the assumption that it is a string.

Either way, we revisited it and noticed that it is preferable to pass the space needed in bytes, so we made a little refactor there.

Addressed in #73303

tyxla commented on client/components/backup-retention-management/retention-options/retention-option-card.tsx:
```diff
@@ -0,0 +1,75 @@
+	};
+
+	return (
+		<div
+			className="retention-option"
+			role="radio"
+			aria-checked={ checked }
+			onClick={ handleOptionClick }
+			onKeyDown={ handleOptionClick }
+			tabIndex={ 0 }
+		>
```
Any more accessible way to compose a radio button? We have a bunch of radio control components available for use.

tyxla commented on client/components/backup-retention-management/retention-options/retention-options-control.tsx:
```diff
@@ -0,0 +1,42 @@
+
+	return (
+		<div className="retention-form__options">
+			{ Object.values( retentionOptions ).map( ( option ) => (
```
If we're going to use `retentionOptions` as an array of values, should we reform the component to expect an array and use that instead of calling `Object.values()`  on an object every time?

Initsogar replied:
Agree, it makes sense. We just prepared a refactor for this.

Addressed in #73303

tyxla commented on client/components/backup-retention-management/index.tsx:
```diff
@@ -0,0 +1,239 @@
+		} else if ( updateRetentionRequestStatus === BACKUP_RETENTION_UPDATE_REQUEST.FAILED ) {
+			setConfirmationDialogVisible( false );
+		}
+		// eslint-disable-next-line react-hooks/exhaustive-deps
```
Why is this necessary?

Initsogar replied:
@tyxla in this case, I would like this check to occur only when `updateRetentionRequestStatus` changes. If I don't include this line, the linter will keep expecting to pass all as dependencies.

tyxla replied:
Technically, the rest of the dependencies can also change and that theoretically should re-run the effect. Is there a good reason to seek workarounds for this default behavior?

Initsogar replied:
I think I see now what's the issue and how it should be.

After updating the retention period, I'm setting the `retentionSelected` at that moment as the `currentRetentionPlan`, which is not the best approach. The right way should be to update the retentionPlan on the redux state using an action dispatcher in the [onSuccess callback](https://github.com/Automattic/wp-calypso/pull/73035/files#diff-d9ef76a91dcc2d8e7451ccddfbc3fc6b732bba3c09dc25de5f55da431b361c2fR24).

I'll take a double look and handle it in a new PR.

Initsogar replied:
Just [submitted a PR](https://github.com/Automattic/wp-calypso/pull/73300) for review related to this, it would be great if you can take a look :) https://github.com/Automattic/wp-calypso/pull/73300

tyxla commented on client/components/backup-retention-management/index.tsx:
```diff
@@ -0,0 +1,239 @@
+	const currentSiteSizeText = useStorageText( estimatedCurrentSiteSize );
+	const storageLimitText = useStorageText( storageLimitBytes );
+
+	const retentionOptionsCards: Record< number, RetentionRadioOptionType > = {
+		[ RETENTION_OPTIONS.RETENTION_DAYS_7 ]: usePrepareRetentionOptions(
+			translate( '7 days' ),
+			RETENTION_OPTIONS.RETENTION_DAYS_7,
+			currentRetentionPlan,
+			retentionSelected === RETENTION_OPTIONS.RETENTION_DAYS_7
+		),
+		[ RETENTION_OPTIONS.RETENTION_DAYS_30 ]: usePrepareRetentionOptions(
+			translate( '30 days' ),
+			RETENTION_OPTIONS.RETENTION_DAYS_30,
+			currentRetentionPlan,
+			retentionSelected === RETENTION_OPTIONS.RETENTION_DAYS_30
+		),
+		[ RETENTION_OPTIONS.RETENTION_DAYS_120 ]: usePrepareRetentionOptions(
+			translate( '120 days' ),
+			RETENTION_OPTIONS.RETENTION_DAYS_120,
+			currentRetentionPlan,
+			retentionSelected === RETENTION_OPTIONS.RETENTION_DAYS_120
+		),
+		[ RETENTION_OPTIONS.RETENTION_DAYS_365 ]: usePrepareRetentionOptions(
+			translate( '1 year' ),
+			RETENTION_OPTIONS.RETENTION_DAYS_365,
+			currentRetentionPlan,
+			retentionSelected === RETENTION_OPTIONS.RETENTION_DAYS_365
+		),
+	};
```
Do we really need to redeclare those on every render?

Initsogar replied:
I think it makes more sense to have it as a component state. We just refactored it.

Addressed in #73303

tyxla commented on client/components/backup-retention-management/constants.ts:
```diff
@@ -0,0 +1,11 @@
+export const RETENTION_OPTIONS = {
+	RETENTION_DAYS_7: 7,
```
Is there a good reason to have those declared in an object? They seem pretty simple and used only in one place.

Initsogar replied:
Double-checked and no. It was an attempt to declare what are the options that we are going to use. I removed this and instead, added the `RetentionPeriod` from `calypso/state/rewind/retention/types` to validate that we use valid values.

Addressed in #73303

tyxla commented on client/components/backup-retention-management/hooks.tsx:
```diff
@@ -0,0 +1,77 @@
+	currentSiteSizeInBytes: number,
+	desiredRetentionDays: number
+): number {
+	return useMemo( () => {
+		return currentSiteSizeInBytes * desiredRetentionDays;
+	}, [ currentSiteSizeInBytes, desiredRetentionDays ] );
```
Have we evaluated the need for each of those memoizations? What's the performance impact and why do we need them? Can we do without them? Would be good to measure and use that to decide whether we need them.

Initsogar replied:
We did some refactors on the way we handle the retention options, and we needed to stop using hooks, as they must only be called at top level.

Either way, at first sight and given that they are simple math operations and the fact that we are not calling too many times those functions, the memoization is not required for `useEstimatedCurrentSiteSize()` and `useEstimateSpaceNeeded()`.

Addressed in #73303

tyxla reviewed (commented):
Hey folks! This is a pretty big PR, have you considered splitting it into smaller pieces? Usually, we recommend such bigger PRs to land in smaller places because that allows us to thoroughly test and review, and ensure that we don't introduce regressions. The more the code, the more difficult it is to catch a bug or regression.

Furthermore, I've left some drive-by feedback and questions.

Please let me know if you're interested in me reviewing a particular aspect of the PR.

Thanks 🙌

Initsogar committed:
Import only FunctionComponent and not everything

Also refactored to use useState and useRef from @wordpress/element instead of React.

Initsogar committed:
Import FunctionComponent in RetentionOptionCard

Initsogar commented:
> Hey folks! This is a pretty big PR, have you considered splitting it into smaller pieces? Usually, we recommend such bigger PRs to land in smaller places because that allows us to thoroughly test and review, and ensure that we don't introduce regressions. The more the code, the more difficult it is to catch a bug or regression.
> 
> Furthermore, I've left some drive-by feedback and questions.

@tyxla thanks for all this valuable feedback! Indeed, it has been quite complicated to make this PR smaller. We tried to add some dependencies in separate PRs in #73031, #73108 and #73007, but I agree that we could do better.

As we need to continue iterating on this new component, I would appreciate it if we could merge this PR and address the comments and suggestion in a new PR. What do you think?

tyxla commented:
> @tyxla thanks for all this valuable feedback! Indeed, it has been quite complicated to make this PR smaller. We tried to add some dependencies in separate PRs in #73031, #73108 and #73007, but I agree that we could do better.
> 
> As we need to continue iterating on this new component, I would appreciate it if we could merge this PR and address the comments and suggestion in a new PR. What do you think?

I really appreciate the effort. Ideally, we should approach changes with the atomic change in mind, and keep large PRs like this only for testing or prototyping. I realize it's hard and time-taking sometimes, but it's worth the effort and our future selves will thank us for that. Anyway, I won't be blocking this PR if you can commit to addressing the feedback in subsequent PRs in the shorter term.

Initsogar commented:
> I really appreciate the effort. Ideally, we should approach changes with the atomic change in mind, and keep large PRs like this only for testing or prototyping. I realize it's hard and time-taking sometimes, but it's worth the effort and our future selves will thank us for that. Anyway, I won't be blocking this PR if you can commit to addressing the feedback in subsequent PRs in the shorter term.

Thank you, @tyxla, I commit to address the feedback in a new PR in short term.

a8ci18n commented:
This Pull Request is now available for translation here: https://translate.wordpress.com/deliverables/7814445

Thank you @Initsogar for including a screenshot in the description! This is really helpful for our translators.

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7814445).

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7814445).

a8ci18n commented:
Translation for this Pull Request [has now been finished](https://translate.wordpress.com/deliverables/overview/7814445).
